dapp_manager/dapp_manager.py

```diff
@@ -88,6 +92,40 @@ def read_file(self, file_type: RunnerFileType, ensure_alive: bool = True) -> str
             self._ensure_alive()
         return self.storage.read_file(file_type)
 
+    def exec_command(self, service: str, command: List[str], timeout: int):
+        self._ensure_alive()
+        m = re.match("^(?P<service_name>.*?)(\\[(?P<idx>\\d+)\\])?$", service)
+        if not m:
+            raise ValueError("`service` parameter format unknown, use ")
+        service_name = m.group("service_name")
+        service_idx = int(m.group("idx") or 0)
+        command_msg = json.dumps({service_name: {service_idx: command}})
+        start = datetime.now()
+        with self.storage.open("data", "r") as data:
+            data.seek(0, 2)
+            self.storage.write_file("commands", command_msg)
+            while datetime.now() < start + timedelta(seconds=timeout):
+                data_out = data.readline()
+                if not data_out:
+                    sleep(COMMAND_OUTPUT_INTERVAL)
+                    continue
+
+                msg = json.loads(data_out)
+
+                if isinstance(msg, dict):
+                    commands = msg.get(service_name, {}).get(str(service_idx))
+                    for cdict in commands:
+                        print(cdict.get("command"))
+                        print("success: ", cdict.get("success"), "\n")
+                        print(cdict.get("stdout"))
+                        stderr = cdict.get("stderr")
+                        if stderr:
+                            print("\nstderr:")
+                            print(stderr)
+                    break
+
+                raise TimeoutError()
```

Comment:
I would split that into more methods for more readability
```python
def exec_command(self, service: str, command: List[str], timeout: int):
     self._ensure_alive()
     ... = eval_input()
     while ... :
         ... = read_msg()
         ... = foo() # last part which IDK if does anything more than `print`
 ```

Reply:
@lucekdudek I updated the function by pulling out the input and output part -> I don't feel it's necessary to split-out the very essence of it...